File: 链表/141.环形链表.py
# 双指针 floyd循环检测算法 龟兔赛跑算法
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow_pointer = head
        fast_pointer = head
        while fast_pointer and fast_pointer.next:
            slow_pointer = slow_pointer.next
            fast_pointer = fast_pointer.next.next
            if slow_pointer == fast_pointer:
                return True
        return False

# 利用不同节点对象的不同(不是值的不同)


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        """使用节点自身当key是独一无二的"""
        memo = set()
        while (head):
            if head in memo:
                return True
            memo.add(head)
            head = curNode.next
        return False

File: 链表/141.环形链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
//* space: O(1)
var hasCycle = function (head) {
  let s = head, f = head;
  // 想要将s移动到之后两个位置, 首先判断s 再判断 s.next
  while (f && f.next) {
    s = s.next;
    f = f.next.next;
    if (s === f) return true;
  }
  return false;
};

//* 通解
var hasCycle = function (head) {
  const memo = new Set();
  while (head) {
    if (memo.has(head)) return true;
    memo.add(head);  //! 判断完将当前节点加入进去
    head = head.next;
  };
  return false;
};
